File: app/src/main/java/art/manguste/android/ArasakaWeapons/OrderAdapter.java
package art.manguste.android.ArasakaWeapons;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.card.MaterialCardView;
import art.manguste.android.ArasakaWeapons.data.Order;
import art.manguste.android.ArasakaWeapons.data.ProductInOrder;

public class OrderAdapter extends RecyclerView.Adapter<OrderAdapter.OrderViewHolder> {

    final private OrderClickListener mOnClickListener;

    public OrderAdapter(OrderClickListener mOnClickListener) {
        this.mOnClickListener = mOnClickListener;
    }

    public interface OrderClickListener{
        void onViewClick(View v, int position, MaterialCardView item, ProductInOrder product);
    }

    @NonNull
    @Override
    public OrderViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        MaterialCardView cv = (MaterialCardView)
                LayoutInflater.from(parent.getContext()).inflate(R.layout.order_card_view, parent, false);

        return new OrderViewHolder(cv);
    }

    @Override
    public void onBindViewHolder(@NonNull OrderViewHolder holder, int position) {
        holder.bind(Order.getCurrentOrder().getProductList().get(position));
    }

    @Override
    public int getItemCount() {
        return Order.getCurrentOrder().getOrderSize();
    }

    /**
     * Holder for a card
     * */
    class OrderViewHolder extends RecyclerView.ViewHolder{
        private MaterialCardView item;
        private ProductInOrder productInOrder;

        /**
         * set data params and listeners
         * */
        public OrderViewHolder(@NonNull MaterialCardView itemView) {
            super(itemView);
            item = itemView;

            ImageButton addButton = itemView.findViewById(R.id.action_decrease_count);
            ImageButton removeButton = itemView.findViewById(R.id.action_increase_count);

            LinearLayout removeProductLayout =  itemView.findViewById(R.id.ll_action_delete_from_cart);
            ImageButton removeProductButton = itemView.findViewById(R.id.action_delete_from_cart);

            // action - decrease items count
            addButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if(mOnClickListener != null)
                        mOnClickListener.onViewClick(v, getAdapterPosition(), item, productInOrder);
                }
            });

            // action - increase items count
            removeButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if(mOnClickListener != null)
                        mOnClickListener.onViewClick(v, getAdapterPosition(), item, productInOrder);
                }
            });

            // action - remove order from cart (linear layout which contains cart icon)
            removeProductLayout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if(mOnClickListener != null)
                        mOnClickListener.onViewClick(v, getAdapterPosition(), item, productInOrder);
                }
            });

            // action - remove order from cart
            removeProductButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if(mOnClickListener != null)
                        mOnClickListener.onViewClick(v, getAdapterPosition(), item, productInOrder);
                }
            });
        }

        /**
         * place product info on a view card
         */
        void bind(ProductInOrder productInOrder){
            this.productInOrder = productInOrder;
            // get data
            int itemsCount = productInOrder.getItemsInOrder();
            String priceString = productInOrder.getProduct().getPriceString();
            String totalPriceString = productInOrder.getProduct().getTotalPriceString(itemsCount);

            //set data
            ((ImageView) item.findViewById(R.id.id_image_in_cart)).setImageResource(productInOrder.getProduct().getImageResourceId());
            ((TextView) item.findViewById(R.id.product_name)).setText(productInOrder.getProduct().getTitle());
            ((TextView) item.findViewById(R.id.tv_items_count)).setText(String.valueOf(itemsCount));

            ((TextView) item.findViewById(R.id.tv_price_card)).setText(priceString);
            ((TextView) item.findViewById(R.id.tv_price_card_total)).setText(totalPriceString);
        }
    }
}

File: app/src/main/java/art/manguste/android/ArasakaWeapons/OrderAdapter.kt
package art.manguste.android.ArasakaWeapons

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import art.manguste.android.ArasakaWeapons.OrderAdapter.OrderViewHolder
import art.manguste.android.ArasakaWeapons.data.Order
import art.manguste.android.ArasakaWeapons.data.ProductInOrder
import com.google.android.material.card.MaterialCardView

/**
 * Adapter for order activity. Shows products in the cart.
 */
class OrderAdapter(private val mOnClickListener: OrderClickListener?) : RecyclerView.Adapter<OrderViewHolder>() {
    interface OrderClickListener {
        fun onViewClick(v: View, position: Int, item: MaterialCardView, product: ProductInOrder?)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): OrderViewHolder {
        val cv = LayoutInflater.from(parent.context).inflate(R.layout.order_card_view, parent, false) as MaterialCardView
        return OrderViewHolder(cv)
    }

    override fun onBindViewHolder(holder: OrderViewHolder, position: Int) {
        Order.productList?.get(position)?.let { holder.bind(it) }
    }

    override fun getItemCount(): Int {
        return Order.orderSize
    }

    /**
     * Holder for a card
     */
    inner class OrderViewHolder(private val item: MaterialCardView) : RecyclerView.ViewHolder(item) {
        private var productInOrder: ProductInOrder? = null

        /**
         * place product info on a view card
         */
        fun bind(productInOrder: ProductInOrder) {
            this.productInOrder = productInOrder
            // get data
            val itemsCount = productInOrder.itemsInOrder
            val priceString = productInOrder.product.priceString
            val totalPriceString = productInOrder.product.getTotalPriceString(itemsCount)

            //set data
            (item.findViewById<View>(R.id.id_image_in_cart) as ImageView).setImageResource(productInOrder.product.imageResourceId)
            (item.findViewById<View>(R.id.productName) as TextView).text = productInOrder.product.title
            (item.findViewById<View>(R.id.count) as TextView).text = itemsCount.toString()
            (item.findViewById<View>(R.id.tv_price_card) as TextView).text = priceString
            (item.findViewById<View>(R.id.tv_price_card_total) as TextView).text = totalPriceString
        }

        /**
         * set data params and listeners
         */
        init {
            val addButton = itemView.findViewById<ImageButton>(R.id.actionDecreaseCount)
            val removeButton = itemView.findViewById<ImageButton>(R.id.actionIncreaseCount)
            val removeProductLayout = itemView.findViewById<LinearLayout>(R.id.ll_action_delete_from_cart)
            val removeProductButton = itemView.findViewById<ImageButton>(R.id.action_delete_from_cart)

            // action - decrease items count
            addButton.setOnClickListener { v -> mOnClickListener?.onViewClick(v, adapterPosition, item, productInOrder) }

            // action - increase items count
            removeButton.setOnClickListener { v -> mOnClickListener?.onViewClick(v, adapterPosition, item, productInOrder) }

            // action - remove order from cart (linear layout which contains cart icon)
            removeProductLayout.setOnClickListener { v -> mOnClickListener?.onViewClick(v, adapterPosition, item, productInOrder) }

            // action - remove order from cart
            removeProductButton.setOnClickListener { v -> mOnClickListener?.onViewClick(v, adapterPosition, item, productInOrder) }
        }
    }
}